src/sky/SkyColoring.cpp

```diff
@@ -0,0 +1,369 @@
+
+    // FIXME: Find proper bounds. Using the world mesh would be okay here.
+    bs::AABox worldBBox = bs::AABox(bs::Vector3(0, 0, 0), bs::Vector3(100, 100, 100));
+
+    // Note: These are some magic values to match what gothic does
+    float fogMidrange = farPlane * 0.4f;
+    float fogMidDelta = farPlane - fogMidrange;
+
+    // Do heightfog-approximation
+    float ytotal  = worldBBox.getMax().y - worldBBox.getMin().y;
+    float fogMinY = worldBBox.getMin().y + 0.5f * ytotal;
+    float fogMaxY = worldBBox.getMin().y + 0.7f * ytotal;
+
+    // Skale fog back depending on how high the camera is
+    float fogScale = fogMaxY - fogMinY != 0 ? (cameraPosition.y - fogMinY) / (fogMaxY - fogMinY) : 0;
+
+    fogScale = bs::Math::clamp(fogScale, 0.0f, 1.0f);
+
+    // Fog should be at least our set distance
+    fogScale = bs::Math::max(mMasterState.fogDistance, fogScale);
+
+    farFog = fogMidrange + (1.0f - fogScale) * fogMidDelta;
+
+    // Apply some user value // FIXME: This should have a getter/setter and all that stuff
+    const float userFogScale = 1.0f;
+    farFog *= userFogScale;
+    nearFog = farFog * 0.3f;
+
+    // REGoth - specific: Let the fog be a little closer because of the long view-distances
+    nearFog *= 0.5f;
+
+    farFog *= 0.4f;
+    nearFog *= 0.4f;
+
+    // Fix up the fog color. The fog should get less intense with decrasing fogFar
+
+    // Compute intensity based on the ZenGins color-base
+    bs::Color base = bs::Color(0.299f, 0.587f, 0.114f);
+
+    // Original engine uses only the red component here as well. Who knows why.
+    bs::Color baseColor =
+        bs::Color(mMasterState.fogColor.r, mMasterState.fogColor.r, mMasterState.fogColor.r);
+
+    // Calculate intensity
+    float intensityValue = std::min(
+        baseColor.r * base.r + baseColor.g * base.g + baseColor.b * base.b, 120.0f / 255.0f);
+
+    bs::Color intensity  = bs::Color(intensityValue, intensityValue, intensityValue);
+    float intensityScale = fogScale * 0.5f;
+
+    // Calculate actual fog color
+    finalFogColor = (1.0f - intensityScale) * mMasterState.fogColor + intensityScale * intensity;
+  }
+
+  void SkyColoring::onWorldNameChanged(const bs::String& newWorldName)
+  {
+    // Must come before loading the config since fillSkyStates initializes everything with defaults
+    fillSkyStates();
+
+    // loadSkyConfig(); // TODO: Needs to be reimplemented, thus commented out
+    // reloadAllSkyTextures(newWorldName); // TODO: Needs to be reimplemented, thus commented out
+  }
+
+  bool SkyColoring::isNightTime() const
+  {
+    return mMasterState.time >= TIME_KEY_1 && mMasterState.time <= TIME_KEY_7;
+  }
+
+  bs::Color SkyColoring::getPolyCloudsLayerColor()
+  {
+    bs::Color color;
+
+    if (isNightTime())
+      color = bs::Color::White;
+    else
+      color = mMasterState.domeColorUpper;
+
+    if (mMasterState.time >= TIME_KEY_3 && mMasterState.time <= TIME_KEY_5)
```

Comment:
That this handles the late night and midnight states (without evening and dawn)

Reply:
I refactored the code. But I don't see a point into making a different component here.